File: services/interactions/src/common/dto/paginated-response.dto.ts
export class PaginatedResponseDto<T> {
  items: T[];
  limit: number;
  offset: number;
  hasMore: boolean;
  totalItems: number;

  constructor(items: T[], total: number, limit: number, offset: number) {
    this.items = items;
    this.totalItems = total;
    this.limit = limit;
    this.offset = offset;
    this.hasMore = offset + items.length < total;
  }
}

File: services/users/src/main/scala/presentation/http/dto/response/query/PaginatedResponseDTO.scala
package presentation.http.dto.response.query

import io.circe.Encoder
import io.circe.generic.semiauto._

case class PaginatedResponseDTO[T](
    data: Seq[T],
    limit: Int,
    offset: Int,
    hasMore: Boolean
)

object PaginatedResponse:
  given [T: Encoder]: Encoder[PaginatedResponseDTO[T]] =
    deriveEncoder[PaginatedResponseDTO[T]].mapJson(_.dropNullValues)
